README.md
# github-repo-stats

A GitHub Action ([in Marketplace](https://github.com/marketplace/actions/github-repo-stats)) built to overcome the [14-day limitation](https://github.com/isaacs/github/issues/399) of GitHub's built-in traffic statistics.

Data that you don't persist today will be gone in two weeks from now.

High-level method description:

* This GitHub Action runs once per day. Each run yields a snapshot of repository traffic statistics (influenced by the past 14 days). Snapshots are persisted via git.
* Each run performs data analysis on all individual snapshots and generates a report from the aggregate — covering an *arbitrarily* long time frame.

## Demo

**Demo 1**:
* [HTML report](https://jgehrcke.github.io/ghrs-test/jgehrcke/github-repo-stats/latest-report/report.html), [PDF report](https://jgehrcke.github.io/ghrs-test/jgehrcke/github-repo-stats/latest-report/report.pdf)
* [Workflow file](https://github.com/jgehrcke/ghrs-test/blob/github-repo-stats/.github/workflows/github-repo-stats.yml), [data branch](https://github.com/jgehrcke/ghrs-test/tree/github-repo-stats/jgehrcke/github-repo-stats)

**Demo 2**:
* [HTML report](https://jgehrcke.github.io/ghrs-test/jgehrcke/covid-19-germany-gae/latest-report/report.html), [PDF report](https://jgehrcke.github.io/ghrs-test/jgehrcke/covid-19-germany-gae/latest-report/report.pdf)
* [Workflow file](https://github.com/jgehrcke/ghrs-test/blob/covid-19-germany-gae/.github/workflows/github-repo-stats.yml), [data branch](https://github.com/jgehrcke/ghrs-test/tree/covid-19-germany-gae/jgehrcke/github-repo-stats)

## Highlights

* The report is generated in two document formats: HTML and PDF.
* The HTML report resembles how GitHub renders Markdown and is meant to be exposed via GitHub pages.
* Charts are based on [Altair](https://github.com/altair-viz/altair)/[Vega](https://vega.github.io/vega/).
* The PDF report contains vector graphics.
* Data updates, aggregation results, and report files are stored in the git repository that you install this Action in: this Action commits changes to a special branch. No cloud storage or database needed. As a result, you have complete and transparent history for data updates and reports, with clear commit messages, in a single place.
* The observed repository (the one to build the report for) can be different from the repository you install this Action in.
* The HTML report can be served right away via GitHub pages (that is how the demo above works).
* Careful data analysis: there are a number of traps ([example](https://github.com/jgehrcke/github-repo-stats/blob/5fefc527288995e2e7e35593db496451580f51db/analyze.py#L748)) when aggregating data based on what the GitHub Traffic API returns. This project tries to not fall for them. One goal of this project is to perform [advanced analysis](https://github.com/jgehrcke/github-repo-stats/blob/5fefc527288995e2e7e35593db496451580f51db/analyze.py#L478) where possible.

## Report content

* Traffic stats:
  * Unique and total views per day
  * Unique and total clones per day
  * Top referrers (where people come from when they land in your repository)
  * Top paths (what people like to look at in your repository)
* Evolution of stargazers
* Evolution of forks

## Credits

This walks on the shoulders of giants:

* [Pandoc](https://pandoc.org/) for rendering HTML from Markdown.
* [Altair](https://altair-viz.github.io/) and [Vega-Lite](https://vega.github.io/vega-lite/) for visualization.
* [Pandas](https://pandas.pydata.org/) for data analysis.
* The [CPython](https://www.python.org/) ecosystem which has always been fun for me to build software in.

## Documentation

### Terminology: *stats repository* and *data repository*

Naming is hard :-). Let's define two concepts and their names:

* The *stats repository* is the repository to fetch stats for and to generate the report for.
* The *data repository* is the repository to store data and report files in. This is also the repository where this Action runs in.

Let me know if you can think of better names.

These two repositories can be the same. But they don't have to be :-).

That is, you can for example set up this Action in a private repository but have it observe a public repository.

### Setup (tutorial)

Example scenario:

* stats repository: `bob/nice-project`
* data repository: `bob/private-ghrs-data-repo`

Create a GitHub Actions workflow file in the *data repository* (in the example this is the repo `bob/private-ghrs-data-repo`). Example path: `.github/workflows/repostats-for-nice-project.yml`.

Example workflow file content with code comments:

```yaml
on:
  schedule:
    # Run this once per day, towards the end of the day for keeping the most
    # recent data point most meaningful (hours are interpreted in UTC).
    - cron: "0 23 * * *"
  workflow_dispatch: # Allow for running this manually.

jobs:
  j1:
    name: repostats-for-nice-project
    runs-on: ubuntu-latest
    steps:
      - name: run-ghrs
        uses: jgehrcke/github-repo-stats@v1.3.0
        with:
          # Define the stats repository (the repo to fetch
          # stats for and to generate the report for).
          # Remove the parameter when the stats repository
          # and the data repository are the same.
          repository: bob/nice-project
          # Set a GitHub API token that can read the stats
          # repository, and that can push to the data
          # repository (which this workflow file lives in),
          # to store data and the report files. This is
          # not needed when the stats repository and the
          # data repository are the same.
          ghtoken: ${{ secrets.ghrs_github_api_token }}

```

**Note:** the recommended way to run this Action is on a schedule, once per day. Really.

**Note:** if you set `ghtoken: ${{ secrets.ghrs_github_api_token }}` as above then in the _data_ repository (where the action is executed) you need to have a secret defined, with the name `GHRS_GITHUB_API_TOKEN` (of course you can change the name in both places).
The content of the secret needs to be an API token that has the `repo` scope for accessing the _stats_ repository.
You can create such a personal access token under [github.com/settings/tokens](https://github.com/settings/tokens).

### Config parameter reference

In the workflow file you can use a number of configuration parameters. They
are specified and documented in the action.yml file (the reference). Here
is a copy for convenience:

* `repository`: Repository spec (`<owner-or-org>/<reponame>`) for the repository
  to fetch statistics for. Ddefault: `${{ github.repository }}`
* `ghtoken`: GitHub API token for reading repo stats and for interacting with
  the data repo (must be set if repo to fetch stats for is not the data repo).
  Default: `${{ github.token }}`
* `databranch`: Branch to push data to (in the data repo).
  Default: `github-repo-stats`
* `ghpagesprefix`: Set this if the data branch in the data repo is exposed via
  GitHub pages. Must not end with a slash.
  Example: `https://jgehrcke.github.io/ghrs-test`
  Default: none

It is recommended that you create the data branch and delete all files from that branch before setting this Action up in your reposistory, so that this data branch appears as a tidy environment.
You can of course remove files from that branch at any other point in time, too.

### Tracking multiple repositories via `matrix`

The GitHub Actions workflow specification language allows for defining a matrix of different job configurations through the [`jobs.<job_id>.strategy.matrix`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) directive.
This can be used for efficiently tracking multiple stats repositories from within the same data repository.

_Example workflow file:_

```yaml
name: fetch-repository-stats
concurrency: fetch-repository-stats

on:
  schedule:
    - cron: "0 23 * * *"
  workflow_dispatch:

jobs:
  run-ghrs-with-matrix:
    name: repostats-for-nice-projects
    runs-on: ubuntu-latest
    strategy:
      matrix:
        # The repositories to generate reports for.
        statsRepo: ['bob/nice-project', 'alice/also-nice-project']
      # Do not cancel&fail all remaining jobs upon first job failure.
      fail-fast: false
      # Help avoid commit conflicts. Note(JP): this should not be
      # necessary anymore, feedback appreciated
      max-parallel: 1
    steps:
      - name: run-ghrs
        uses: jgehrcke/github-repo-stats@v1.2.0
        with:
          # Repo to fetch stats for and to generate the report for.
          repository: ${{ matrix.statsRepo }}
          # Token that can read the stats repository and that
          # can push to the data repository.
          ghtoken: ${{ secrets.ghrs_github_api_token }}
          # Data branch: Branch to push data to (in the data repo).
          databranch: main
```

## Developer notes

### CLI tests

Here is how to run [bats](https://github.com/bats-core/bats-core)-based checks from within a checkout:

```bash
$ git clone https://github.com/jgehrcke/github-repo-stats
$ cd github-repo-stats/

$ make clitests
...
1..5
ok 1 analyze.py: snapshots: some, vcagg: yes, stars: some, forks: none
ok 2 analyze.py: snapshots: some, vcagg: yes, stars: none, forks: some
ok 3 analyze.py: snapshots: some, vcagg: yes, stars: some, forks: some
ok 4 analyze.py: snapshots: some, vcagg: no, stars: some, forks: some
ok 5 analyze.py + pdf.py: snapshots: some, vcagg: no, stars: some, forks: some
```

### Lint

```bash
$ make lint
...
All done! ✨ 🍰 ✨
...
```

### local run of entrypoint.sh

Set environment variables, example:

```bash
export GITHUB_REPOSITORY=jgehrcke/ghrs-test
export GITHUB_WORKFLOW="localtesting"
export INPUT_DATABRANCH=databranch-test
export INPUT_GHTOKEN="c***1"
export INPUT_REPOSITORY=jgehrcke/covid-19-germany-gae
export INPUT_GHPAGESPREFIX="none"
export GHRS_FILES_ROOT_PATH="/home/jp/dev/github-repo-stats"
export GHRS_TESTING="true"
```

(for an up-to-date list of required env vars see `.github/workflows/prs.yml`)

Run in empty directory. Example:

```bash
cd /tmp/ghrstest
rm -rf .* *; bash /home/jp/dev/github-repo-stats/entrypoint.sh
```

## Further resources

* [“GitHub Stars” -- useful for *what*?](https://opensource.stackexchange.com/questions/5110/github-stars-is-a-very-useful-metric-but-for-what/5114#5114)
* [GitHub Traffic API docs](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#traffic)
* [Do your own views count?](https://stackoverflow.com/a/63697886/145400)


CHANGES.md
# Changelog

## 1.4.0 (2022-05-18)

Bug fixes:

* Do not fail Action when the number of data points to be plotted in a single graph grows beyond 5000  (also see [#52](https://github.com/jgehrcke/github-repo-stats/issues/52)).
* Do not fail Action when no forks exist (also see [#49](https://github.com/jgehrcke/github-repo-stats/issues/49)).

Data handling:

* Make the top-N plots for referrers and pages less busy: show the top 7 instead of top 10 and downsample to one data point per five days if the total number of samples would otherwise grow beyond 3000.

Misc:

* Log output: improved debuggability.
* Conservative dependency update; new base image.

Testing: better coverage of `entrypoint.sh` logic in CI, better debuggability of `bats`-based CLI tests.

## 1.3.0 (2021-12-03)

Thanks for all the feedback and contributions.

Report improvements:

* Automatically use semi-logarithmic plotting when it appears to make sense. This is supposed to help in cases of traffic spikes much higher than the baseline. (With linear plotting, the normal traffic level would then appear to be very close to zero, and variations around it would hardly be visible.)
* Add data point tooltips as mouse hover effect in the HTML report (also see [#28](https://github.com/jgehrcke/github-repo-stats/issues/28)).
* Display cumulative view/clone counts (total count until day of report generation, also see [#31](https://github.com/jgehrcke/github-repo-stats/issues/31)).
* Top referrer/path plots: move the legend so that these plots have the same width as all other plots, making comparison easier.
* Synchronize date axes where it makes sense:
  * The plots for view/clone data now all show the same time window.
  * The plots for star/fork data always show the same time window.
  * The star/fork plots only show the same time window as the view/clone plots when data collection started at the time of or before the first star/fork event. Else, the star/fork plots go further into the past.
* Create a meaningful fork/stargazer section for the special cases of zero forks/stars, respectively (also see [#41](https://github.com/jgehrcke/github-repo-stats/issues/41), [#43](https://github.com/jgehrcke/github-repo-stats/issues/43)).
* Tweak plot style (marker size, tick label angle, and others).

Job robustness and performance:

* Perform a shallow clone of the data repository for faster job execution ([#32](https://github.com/jgehrcke/github-repo-stats/issues/32), [#34](https://github.com/jgehrcke/github-repo-stats/issues/34)).

Bug fixes:

* Fix a path/referrer snapshot time series aggregation bug where a significant fraction of the available data was not visualized ([#36](https://github.com/jgehrcke/github-repo-stats/issues/36)).
* View/clone analysis: do not crash anymore when there is an aggregate file but no new snapshots ([#37](https://github.com/jgehrcke/github-repo-stats/issues/37)).
* Fix `git pull` and `git push` errors for the first-run scenario where the data branch does not yet exist on the data repository (see [#30](https://github.com/jgehrcke/github-repo-stats/issues/30), [#33](https://github.com/jgehrcke/github-repo-stats/issues/33), [#35](https://github.com/jgehrcke/github-repo-stats/issues/35)).

Misc:

* Reduce log verbosity.
* Update several dependencies (for example, use Altair 4.2.0 with Vega-Lite 4.17.0).

**Testing:** important changes coming hand-in-hand with this release are on the testing and continuous integration (CI) front.
Given the growing user base, I did not feel comfortable anymore with the reliance on manual testing.
I introduced a framework for high-level CLI invocation tests with specific data scenarios, now executed by CI as part of every commit.
This will help prevent regressions from happening.
The `README` contains instructions for how to run these tests locally.
I have also introduced more linting, and added a `mypy` check to CI (starting out with very basic typing information in the code base here and there).

## 1.2.0 (2021-09-11)

Data handling:

* Fix an edge case for exclusively empty views/clones fragments ([issue #15](https://github.com/jgehrcke/github-repo-stats/issues/15)).
* Expect an edge case where `fetch.py` exits without having generated new snapshot files: in that case, regenerate the report using the most recent set of data ([issue #17](https://github.com/jgehrcke/github-repo-stats/issues/17)).

Documentation:

* Fix cron syntax in README ([issue #20](https://github.com/jgehrcke/github-repo-stats/issues/20)).
* Document elegant method for a multi-repo workflow using the `matrix` approach ([PR #26](https://github.com/jgehrcke/github-repo-stats/pull/26)). Thanks to David Farrell and to Egil Hansen.

Job robustness and performance:

* Use a pre-built Docker container image ([jgehrcke/github-repo-stats-base](https://hub.docker.com/r/jgehrcke/github-repo-stats-base)) to base this action on. This image includes heavy Python and browser dependencies. This approach significantly reduces the probability for an action run to fail as of one of the many potential transient issues affecting a complex Docker image build. This also significantly reduces the time it takes for completing the `Build container for action use` build step (from ~2 minutes to ~30 seconds). Context: [issue #24](https://github.com/jgehrcke/github-repo-stats/issues/24) and [PR #25](https://github.com/jgehrcke/github-repo-stats/pull/25). If you have security concerns please open an issue and let's talk it through.

## 1.1.0 (2021-04-14)

Job robustness:

* Consolidate data branch writing so that multiple jobs operating concurrently on the same branch are less likely to fail ([issue #9](https://github.com/jgehrcke/github-repo-stats/issues/9)). Thanks to Henry Bley-Vroman and Dmytro Chasovskyi for the feedback.
* Fix an edge case for missing path/referrer data ([issue #8](https://github.com/jgehrcke/github-repo-stats/issues/8)).

fetch.py
#!/usr/bin/env python
# Copyright 2018 - 2020 Dr. Jan-Philip Gehrcke
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

import argparse
import logging
import os
import json
from datetime import datetime

import sys
from typing import Tuple


import pandas as pd
from github import Github, Repository  # type: ignore
import requests
import retrying  # type: ignore
import pytz


"""
prior art
https://github.com/MTG/github-traffic
https://github.com/nchah/github-traffic-stats/
https://github.com/sangonzal/repository-traffic-action

makes use of code and methods from my other projects at
https://github.com/jgehrcke/dcos-dev-prod-analysis
https://github.com/jgehrcke/bouncer-log-analysis
https://github.com/jgehrcke/goeffel
"""


log = logging.getLogger()
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s.%(msecs)03d %(levelname)s:%(threadName)s: %(message)s",
    datefmt="%y%m%d-%H:%M:%S",
)


# Get tz-aware datetime object corresponding to invocation time.
NOW = pytz.timezone("UTC").localize(datetime.utcnow())
INVOCATION_TIME_STRING = NOW.strftime("%Y-%m-%d_%H%M%S")

if not os.environ.get("GHRS_GITHUB_API_TOKEN", None):
    sys.exit("error: environment variable GHRS_GITHUB_API_TOKEN empty or not set")

GHUB = Github(login_or_token=os.environ["GHRS_GITHUB_API_TOKEN"].strip(), per_page=100)


def main() -> None:
    args = parse_args()
    # Full name of repo with slash (including owner/org)
    repo: Repository.Repository = GHUB.get_repo(args.repo)
    log.info("Working with repository `%s`", repo)
    log.info("Request quota limit: %s", GHUB.get_rate_limit())

    (
        df_views_clones,
        df_referrers_snapshot_now,
        df_paths_snapshot_now,
    ) = fetch_all_traffic_api_endpoints(repo)

    outdir_path = args.snapshot_directory
    log.info("current working directory: %s", os.getcwd())
    log.info("write output CSV files to directory: %s", outdir_path)

    if len(df_views_clones):
        df_views_clones.to_csv(
            os.path.join(
                outdir_path,
                f"{INVOCATION_TIME_STRING}_views_clones_series_fragment.csv",
            )
        )
    else:
        log.info("do not write df_views_clones: empty")

    if len(df_referrers_snapshot_now):
        df_referrers_snapshot_now.to_csv(
            os.path.join(
                outdir_path, f"{INVOCATION_TIME_STRING}_top_referrers_snapshot.csv"
            )
        )
    else:
        log.info("do not write df_referrers_snapshot_now: empty")

    if len(df_paths_snapshot_now):
        df_paths_snapshot_now.to_csv(
            os.path.join(
                outdir_path, f"{INVOCATION_TIME_STRING}_top_paths_snapshot.csv"
            )
        )
    else:
        log.info("do not write df_paths_snapshot_now: empty")

    if args.fork_ts_outpath:
        fetch_and_write_fork_ts(repo, args.fork_ts_outpath)

    if args.stargazer_ts_outpath:
        fetch_and_write_stargazer_ts(repo, args.stargazer_ts_outpath)

    log.info("done!")


def fetch_and_write_stargazer_ts(repo: Repository.Repository, path: str):
    dfstarscsv = get_stars_over_time(repo)
    log.info("stars_cumulative, for CSV file:\n%s", dfstarscsv)
    tpath = path + ".tmp"  # todo: rnd string
    log.info(
        "write stargazer time series to %s, then rename to %s",
        tpath,
        path,
    )
    dfstarscsv.to_csv(tpath, index_label="time_iso8601")
    os.rename(tpath, path)


def fetch_and_write_fork_ts(repo: Repository.Repository, path: str):
    dfforkcsv = get_forks_over_time(repo)
    log.info("forks_cumulative, for CSV file:\n%s", dfforkcsv)
    tpath = path + ".tmp"  # todo: rnd string
    log.info(
        "write fork time series to %s, then rename to %s",
        tpath,
        path,
    )
    dfforkcsv.to_csv(tpath, index_label="time_iso8601")
    os.rename(tpath, path)


def fetch_all_traffic_api_endpoints(
    repo,
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:

    log.info("fetch top referrers")
    df_referrers_snapshot_now = referrers_to_df(fetch_top_referrers(repo))

    log.info("fetch top paths")
    df_paths_snapshot_now = paths_to_df(fetch_top_paths(repo))

    log.info("fetch data for clones")
    df_clones = clones_or_views_to_df(fetch_clones(repo), "clones")

    log.info("fetch data for views")
    df_views = clones_or_views_to_df(fetch_views(repo), "views")

    # Note that df_clones and df_views should have the same datetime index, but
    # there is no guarantee for that. Create two separate data frames, then
    # merge / align dynamically.
    if not df_clones.index.equals(df_views.index):
        log.info("special case: df_views and df_clones have different index")
    else:
        log.info("indices of df_views and df_clones are equal")

    log.info("union-merge views and clones")
    # https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html#set-logic-on-the-other-axes
    # Build union of the two data frames. Zero information loss, in case the
    # two indices aree different.
    df_views_clones = pd.concat([df_clones, df_views], axis=1, join="outer")
    log.info("df_views_clones:\n%s", df_views_clones)

    return df_views_clones, df_referrers_snapshot_now, df_paths_snapshot_now


def parse_args():
    parser = argparse.ArgumentParser(
        description="Fetch traffic data for GitHub repository. Requires the "
        "environment variables GITHUB_USERNAME and GITHUB_APITOKEN to be set."
    )

    parser.add_argument(
        "repo",
        metavar="REPOSITORY",
        help="Owner/organization and repository. Must contain a slash. "
        "Example: coke/truck",
    )

    parser.add_argument(
        "--snapshot-directory",
        type=str,
        default="",
        help="Snapshot/fragment directory. Default: _ghrs_{owner}_{repo}",
    )

    parser.add_argument(
        "--fork-ts-outpath",
        default="",
        metavar="PATH",
        help="Fetch fork time series and write to this CSV file. Overwrite if file exists.",
    )

    parser.add_argument(
        "--stargazer-ts-outpath",
        default="",
        metavar="PATH",
        help="Fetch stargazer time series and write to this CSV file. Overwrite if file exists.",
    )

    args = parser.parse_args()

    if "/" not in args.repo:
        sys.exit("missing slash in REPOSITORY spec")

    ownerid, repoid = args.repo.split("/")
    outdir_path_default = f"_ghrs_{ownerid}_{repoid}"

    if not args.snapshot_directory:
        args.snapshot_directory = outdir_path_default

    log.info("processed args: %s", json.dumps(vars(args), indent=2))

    if os.path.exists(args.snapshot_directory):
        if not os.path.isdir(args.snapshot_directory):
            log.error(
                "the specified output directory path does not point to a directory: %s",
                args.snapshot_directory,
            )
            sys.exit(1)

        log.info("output directory already exists: %s", args.snapshot_directory)

    else:
        log.info("create output directory: %s", args.snapshot_directory)
        log.info("absolute path: %s", os.path.abspath(args.snapshot_directory))
        # If there is a race: do not error out.
        os.makedirs(args.snapshot_directory, exist_ok=True)

    return args


def referrers_to_df(top_referrers) -> pd.DataFrame:
    series_referrers = []
    series_views_unique = []
    series_views_total = []
    for p in top_referrers:
        series_referrers.append(p.referrer)
        series_views_total.append(int(p.count))
        series_views_unique.append(int(p.uniques))

    df = pd.DataFrame(
        data={
            "views_total": series_views_total,
            "views_unique": series_views_unique,
        },
        index=series_referrers,
    )
    df.index.name = "referrer"

    # Attach metadata to dataframe, still experimental -- also see
    # https://stackoverflow.com/q/52122674/145400
    df.attrs["snapshot_time"] = NOW.isoformat()
    return df


def paths_to_df(top_paths) -> pd.DataFrame:

    series_url_paths = []
    series_views_unique = []
    series_views_total = []
    for p in top_paths:
        series_url_paths.append(p.path)
        series_views_total.append(int(p.count))
        series_views_unique.append(int(p.uniques))

    df = pd.DataFrame(
        data={
            "views_total": series_views_total,
            "views_unique": series_views_unique,
        },
        index=series_url_paths,
    )
    df.index.name = "url_path"

    # Attach metadata to dataframe, new as of pandas 1.0 -- also see
    # https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.attrs.html
    # https://github.com/pandas-dev/pandas/issues/28283
    # https://stackoverflow.com/q/52122674/145400
    df.attrs["snapshot_time"] = NOW.isoformat()
    return df


def clones_or_views_to_df(items, metric) -> pd.DataFrame:
    assert metric in ["clones", "views"]

    series_count_total = []
    series_count_unique = []
    series_timestamps = []

    for sample in items:
        # GitHub API docs say "Timestamps are aligned to UTC".
        # `sample.timestamp` is a tz-naive datetime object.
        series_timestamps.append(sample.timestamp)
        series_count_total.append(int(sample.count))
        series_count_unique.append(int(sample.uniques))

    # Attach timezone information to `pd.DatetimeIndex` (make this index
    # tz-aware, leave actual numbers intact).
    df = pd.DataFrame(
        data={
            f"{metric}_total": series_count_total,
            f"{metric}_unique": series_count_unique,
        },
        index=pd.DatetimeIndex(data=series_timestamps, tz="UTC"),
    )
    df.index.name = "time_iso8601"

    log.info("built dataframe for %s:\n%s", metric, df)
    log.info("dataframe datetimeindex detail: %s", df.index)
    return df


def get_forks_over_time(repo: Repository.Repository) -> pd.DataFrame:
    # TODO: for ~10k forks repositories, this operation is too costly for doing
    # it as part of each analyzer invocation. Move this to the fetcher, and
    # persist the data.
    log.info("fetch fork time series for repo %s", repo)

    reqlimit_before = GHUB.get_rate_limit().core.remaining
    log.info("GH request limit before operation: %s", reqlimit_before)

    forks = []
    for count, fork in enumerate(repo.get_forks(), 1):
        # Store `PullRequest` object with integer key in dictionary.
        forks.append(fork)
        if count % 200 == 0:
            log.info("%s forks fetched", count)

    reqlimit_after = GHUB.get_rate_limit().core.remaining
    log.info("GH request limit after operation: %s", reqlimit_after)
    log.info("http requests made (approximately): %s", reqlimit_before - reqlimit_after)
    log.info("current fork count: %s", len(forks))

    # The GitHub API returns ISO 8601 timestamp strings encoding the timezone
    # via the Z suffix, i.e. Zulu time, i.e. UTC. pygithub doesn't parse that
    # timezone. That is, whereas the API returns `starred_at` in UTC, the
    # datetime obj created by pygithub is a naive one. Correct for that.
    forktimes_aware = [pytz.timezone("UTC").localize(f.created_at) for f in forks]

    # Create sorted pandas DatetimeIndex
    dtidx = pd.to_datetime(forktimes_aware)
    dtidx = dtidx.sort_values()

    # Each timestamp corresponds to *1* fork event. Build cumulative sum over
    # time.
    df = pd.DataFrame(
        data={"fork_events": [1] * len(forks)},
        index=dtidx,
    )
    df.index.name = "time"
    df["forks_cumulative"] = df["fork_events"].cumsum()
    df = df.drop(columns=["fork_events"]).astype(int)
    log.info("forks df: \n%s", df)
    return df


def get_stars_over_time(repo: Repository.Repository) -> pd.DataFrame:
    # TODO: for ~10k stars repositories, this operation is too costly for doing
    # it as part of each analyzer invocation. Move this to the fetcher, and
    # persist the data.
    log.info("fetch stargazer time series for repo %s", repo)

    reqlimit_before = GHUB.get_rate_limit().core.remaining

    log.info("GH request limit before fetch operation: %s", reqlimit_before)

    gazers = []

    # TODO for addressing the 10ks challenge: save state to disk, and refresh
    # using reverse order iteration. See for repo in user.get_repos().reversed
    for count, gazer in enumerate(repo.get_stargazers_with_dates(), 1):
        # Store `PullRequest` object with integer key in dictionary.
        gazers.append(gazer)
        if count % 200 == 0:
            log.info("%s gazers fetched", count)

    reqlimit_after = GHUB.get_rate_limit().core.remaining
    log.info("GH request limit after fetch operation: %s", reqlimit_after)
    log.info("http requests made (approximately): %s", reqlimit_before - reqlimit_after)
    log.info("stargazer count: %s", len(gazers))

    # The GitHub API returns ISO 8601 timestamp strings encoding the timezone
    # via the Z suffix, i.e. Zulu time, i.e. UTC. pygithub doesn't parze that
    # timezone. That is, whereas the API returns `starred_at` in UTC, the
    # datetime obj created by pygithub is a naive one. Correct for that.
    startimes_aware = [pytz.timezone("UTC").localize(g.starred_at) for g in gazers]

    # Work towards a dataframe of the following shape:
    #                            star_events  stars_cumulative
    # time
    # 2020-11-26 16:25:37+00:00            1                 1
    # 2020-11-26 16:27:23+00:00            1                 2
    # 2020-11-26 16:30:05+00:00            1                 3
    # 2020-11-26 17:31:57+00:00            1                 4
    # 2020-11-26 17:48:48+00:00            1                 5
    # ...                                ...               ...
    # 2020-12-19 19:48:58+00:00            1               327
    # 2020-12-22 04:44:35+00:00            1               328
    # 2020-12-22 19:00:42+00:00            1               329
    # 2020-12-25 05:01:42+00:00            1               330
    # 2020-12-28 01:07:55+00:00            1               331

    # Create sorted pandas DatetimeIndex
    dtidx = pd.to_datetime(startimes_aware)
    dtidx = dtidx.sort_values()

    # Each timestamp corresponds to *1* star event. Build cumulative sum over
    # time.
    df = pd.DataFrame(
        data={"star_events": [1] * len(gazers)},
        index=dtidx,
    )
    df.index.name = "time"

    df["stars_cumulative"] = df["star_events"].cumsum()
    df = df.drop(columns=["star_events"]).astype(int)
    log.info("stargazer df\n %s", df)
    return df


def handle_rate_limit_error(exc):

    if "wait a few minutes before you try again" in str(exc):
        log.warning("GitHub abuse mechanism triggered, wait 60 s, retry")
        return True

    if "403" in str(exc):
        if "Resource not accessible by integration" in str(exc):
            log.error(
                'this appears to be a permanent error, as in "access denied -- do not retry": %s',
                str(exc),
            )
            sys.exit(1)

        log.warning("Exception contains 403, wait 60 s, retry: %s", str(exc))
        # The request count quota is not necessarily responsible for this
        # exception, but it usually is. Log the expected local time when the
        # new quota arrives.
        unix_timestamp_quota_reset = GHUB.rate_limiting_resettime
        local_time = datetime.fromtimestamp(unix_timestamp_quota_reset)
        log.info("New req count quota at: %s", local_time.strftime("%Y-%m-%d %H:%M:%S"))
        return True

    # For example, `RemoteDisconnected` is a case I have seen in production.
    if isinstance(exc, requests.exceptions.RequestException):
        log.warning("RequestException, wait 60 s, retry: %s", str(exc))
        return True

    return False


@retrying.retry(wait_fixed=60000, retry_on_exception=handle_rate_limit_error)
def fetch_clones(repo):
    clones = repo.get_clones_traffic()
    return clones["clones"]


@retrying.retry(wait_fixed=60000, retry_on_exception=handle_rate_limit_error)
def fetch_views(repo):
    views = repo.get_views_traffic()
    return views["views"]


@retrying.retry(wait_fixed=60000, retry_on_exception=handle_rate_limit_error)
def fetch_top_referrers(repo):
    return repo.get_top_referrers()


@retrying.retry(wait_fixed=60000, retry_on_exception=handle_rate_limit_error)
def fetch_top_paths(repo):
    return repo.get_top_paths()


if __name__ == "__main__":
    main()
